File: ios/Classes/Bridge/AdsSettingsMethodHandler.swift
//
//  AdsSettingsMethodHandler.swift
//  clever_ads_solutions
//
//  Copyright © 2024 CleverAdsSolutions LTD, CAS.AI. All rights reserved.
//

import CleverAdsSolutions
import Flutter

private let channelName = "com.cleveradssolutions.plugin.flutter/ads_settings"

class AdsSettingsMethodHandler: MethodHandler {
    init() {
        super.init(channelName: channelName)
    }

    override func onMethodCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        switch call.method {
        case "getTaggedAudience": getTaggedAudience(result)
        case "setTaggedAudience": setTaggedAudience(call, result)
        case "getUserConsent": getUserConsent(result)
        case "setUserConsent": setUserConsent(call, result)
        case "getVendorConsent": getVendorConsent(call, result)
        case "getAdditionalConsent": getAdditionalConsent(call, result)
        case "getCPPAStatus": getCPPAStatus(result)
        case "setCCPAStatus": setCCPAStatus(call, result)
        case "getMutedAdSounds": getMutedAdSounds(result)
        case "setMutedAdSounds": setMutedAdSounds(call, result)
        case "getDebugMode": getDebugMode(result)
        case "setDebugMode": setDebugMode(call, result)
        case "addTestDeviceId": setTestDeviceId(call, result)
        case "setTestDeviceId": setTestDeviceId(call, result)
        case "setTestDeviceIds": setTestDeviceIds(call, result)
        case "clearTestDeviceIds": clearTestDeviceIds(result)
        case "getTrialAdFreeInterval": getTrialAdFreeInterval(result)
        case "setTrialAdFreeInterval": setTrialAdFreeInterval(call, result)
        case "getBannerRefreshDelay": getBannerRefreshDelay(result)
        case "setBannerRefreshDelay": setBannerRefreshDelay(call, result)
        case "getInterstitialInterval": getInterstitialInterval(result)
        case "setInterstitialInterval": setInterstitialInterval(call, result)
        case "restartInterstitialInterval": restartInterstitialInterval(result)
        case "isAllowInterstitialAdsWhenVideoCostAreLower": isAllowInterstitialAdsWhenVideoCostAreLower(result)
        case "allowInterstitialAdsWhenVideoCostAreLower": allowInterstitialAdsWhenVideoCostAreLower(call, result)
        case "getLoadingMode": getLoadingMode(result)
        case "setLoadingMode": setLoadingMode(call, result)
        default: super.onMethodCall(call, result)
        }
    }

    private func getTaggedAudience(_ result: @escaping FlutterResult) {
        result(CAS.settings.taggedAudience.rawValue)
    }

    private func setTaggedAudience(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("taggedAudience", result) { index in
            if let value = CASAudience(rawValue: index) {
                CAS.settings.taggedAudience = value
            }
        }
    }

    private func getUserConsent(_ result: @escaping FlutterResult) {
        result(CAS.settings.userConsent.rawValue)
    }

    private func setUserConsent(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("userConsent", result) { index in
            if let value = CASConsentStatus(rawValue: index) {
                CAS.settings.userConsent = value
            }
        }
    }

    private func getVendorConsent(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturnResult("vendorId", result) { vendorId in
            CAS.settings.getVendorConsent(vendorId: vendorId).rawValue
        }
    }

    private func getAdditionalConsent(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturnResult("providerId", result) { providerId in
            CAS.settings.getAdditionalConsent(providerId: providerId).rawValue
        }
    }

    private func getCPPAStatus(_ result: @escaping FlutterResult) {
        result(CAS.settings.userCCPAStatus.rawValue)
    }

    private func setCCPAStatus(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("ccpa", result) { index in
            if let value = CASCCPAStatus(rawValue: index) {
                CAS.settings.userCCPAStatus = value
            }
        }
    }

    private func getMutedAdSounds(_ result: @escaping FlutterResult) {
        result(CAS.settings.mutedAdSounds)
    }

    private func setMutedAdSounds(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("muted", result) { value in
            CAS.settings.mutedAdSounds = value
        }
    }

    private func getDebugMode(_ result: @escaping FlutterResult) {
        result(CAS.settings.debugMode)
    }

    private func setDebugMode(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("enable", result) { value in
            CAS.settings.debugMode = value
        }
    }

    private func setTestDeviceId(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("deviceId", result) { deviceId in
            CAS.settings.setTestDevice(ids: [deviceId])
        }
    }

    private func setTestDeviceIds(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("deviceIds", result) { deviceIds in
            CAS.settings.setTestDevice(ids: deviceIds)
        }
    }

    private func clearTestDeviceIds(_ result: @escaping FlutterResult) {
        CAS.settings.setTestDevice(ids: [])
        result(nil)
    }

    private func getTrialAdFreeInterval(_ result: @escaping FlutterResult) {
        result(CAS.settings.trialAdFreeInterval)
    }

    private func setTrialAdFreeInterval(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("interval", result) { interval in
            CAS.settings.trialAdFreeInterval = interval
        }
    }

    private func getBannerRefreshDelay(_ result: @escaping FlutterResult) {
        result(CAS.settings.bannerRefreshInterval)
    }

    private func setBannerRefreshDelay(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("delay", result) { delay in
            CAS.settings.bannerRefreshInterval = delay
        }
    }

    private func getInterstitialInterval(_ result: @escaping FlutterResult) {
        result(CAS.settings.interstitialInterval)
    }

    private func setInterstitialInterval(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("delay", result) { delay in
            CAS.settings.interstitialInterval = delay
        }
    }

    private func restartInterstitialInterval(_ result: @escaping FlutterResult) {
        CAS.settings.restartInterstitialInterval()
        result(nil)
    }

    private func isAllowInterstitialAdsWhenVideoCostAreLower(_ result: @escaping FlutterResult) {
        result(CAS.settings.isInterstitialAdsWhenVideoCostAreLowerAllowed())
    }

    private func allowInterstitialAdsWhenVideoCostAreLower(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("enable", result) { enable in
            CAS.settings.setInterstitialAdsWhenVideoCostAreLower(allow: enable)
        }
    }

    private func getLoadingMode(_ result: @escaping FlutterResult) {
        result(CAS.settings.getLoadingMode())
    }

    private func setLoadingMode(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("loadingMode", result) { loadingMode in
            CAS.settings.setLoading(mode: loadingMode)
        }
    }
}

File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/bridge/AdsSettingsMethodHandler.kt
package com.cleveradssolutions.plugin.flutter.bridge

import com.cleveradssolutions.plugin.flutter.bridge.base.MethodHandler
import com.cleveradssolutions.plugin.flutter.util.getArgAndReturn
import com.cleveradssolutions.plugin.flutter.util.getArgAndReturnResult
import com.cleveradssolutions.plugin.flutter.util.success
import com.cleversolutions.ads.android.CAS
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

private const val CHANNEL_NAME = "cleveradssolutions/ads_settings"

class AdsSettingsMethodHandler(
    binding: FlutterPluginBinding
) : MethodHandler(binding, CHANNEL_NAME) {

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getTaggedAudience" -> getTaggedAudience(result)
            "setTaggedAudience" -> setTaggedAudience(call, result)
            "getUserConsent" -> getUserConsent(result)
            "setUserConsent" -> setUserConsent(call, result)
            "getVendorConsent" -> getVendorConsent(call, result)
            "getAdditionalConsent" -> getAdditionalConsent(call, result)
            "getCPPAStatus" -> getCPPAStatus(result)
            "setCCPAStatus" -> setCCPAStatus(call, result)
            "getMutedAdSounds" -> getMutedAdSounds(result)
            "setMutedAdSounds" -> setMutedAdSounds(call, result)
            "getDebugMode" -> getDebugMode(result)
            "setDebugMode" -> setDebugMode(call, result)
            "addTestDeviceId" -> addTestDeviceId(call, result)
            "setTestDeviceId" -> setTestDeviceId(call, result)
            "setTestDeviceIds" -> setTestDeviceIds(call, result)
            "getTrialAdFreeInterval" -> getTrialAdFreeInterval(result)
            "setTrialAdFreeInterval" -> setTrialAdFreeInterval(call, result)
            "getBannerRefreshDelay" -> getBannerRefreshDelay(result)
            "setBannerRefreshDelay" -> setBannerRefreshDelay(call, result)
            "getInterstitialInterval" -> getInterstitialInterval(result)
            "setInterstitialInterval" -> setInterstitialInterval(call, result)
            "restartInterstitialInterval" -> restartInterstitialInterval(result)
            "isAllowInterstitialAdsWhenVideoCostAreLower" -> isAllowInterstitialAdsWhenVideoCostAreLower(
                result
            )

            "allowInterstitialAdsWhenVideoCostAreLower" -> allowInterstitialAdsWhenVideoCostAreLower(
                call,
                result
            )

            "getLoadingMode" -> getLoadingMode(result)
            "setLoadingMode" -> setLoadingMode(call, result)
            else -> super.onMethodCall(call, result)
        }
    }

    private fun getTaggedAudience(result: MethodChannel.Result) {
        result.success(CAS.settings.taggedAudience)
    }

    private fun setTaggedAudience(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<Int>("taggedAudience", result) {
            CAS.settings.taggedAudience = it
        }
    }

    private fun getUserConsent(result: MethodChannel.Result) {
        result.success(CAS.settings.userConsent)
    }

    private fun setUserConsent(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<Int>("userConsent", result) {
            CAS.settings.userConsent = it
        }
    }

    private fun getVendorConsent(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturnResult<Int>("vendorId", result) {
            CAS.settings.getVendorConsent(it)
        }
    }

    private fun getAdditionalConsent(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturnResult<Int>("providerId", result) {
            CAS.settings.getAdditionalConsent(it)
        }
    }

    private fun getCPPAStatus(result: MethodChannel.Result) {
        result.success(CAS.settings.ccpaStatus)
    }

    private fun setCCPAStatus(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<Int>("ccpa", result) {
            CAS.settings.ccpaStatus = it
        }
    }

    private fun getMutedAdSounds(result: MethodChannel.Result) {
        result.success(CAS.settings.mutedAdSounds)
    }

    private fun setMutedAdSounds(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<Boolean>("muted", result) {
            CAS.settings.mutedAdSounds = it
        }
    }

    private fun getDebugMode(result: MethodChannel.Result) {
        result.success(CAS.settings.debugMode)
    }

    private fun setDebugMode(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<Boolean>("enable", result) {
            CAS.settings.debugMode = it
        }
    }

    private fun addTestDeviceId(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<String>("deviceId", result) {
            CAS.settings.testDeviceIDs += it
        }
    }

    private fun setTestDeviceId(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<String>("deviceId", result) {
            CAS.settings.testDeviceIDs = setOf(it)
        }
    }

    private fun setTestDeviceIds(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<Set<String>>("deviceIds", result) {
            CAS.settings.testDeviceIDs = it
        }
    }

    private fun getTrialAdFreeInterval(result: MethodChannel.Result) {
        result.success(CAS.settings.trialAdFreeInterval)
    }

    private fun setTrialAdFreeInterval(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<Int>("interval", result) {
            CAS.settings.bannerRefreshInterval = it
        }
    }

    private fun getBannerRefreshDelay(result: MethodChannel.Result) {
        result.success(CAS.settings.bannerRefreshInterval)
    }

    private fun setBannerRefreshDelay(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<Int>("delay", result) {
            CAS.settings.bannerRefreshInterval = it
        }
    }

    private fun getInterstitialInterval(result: MethodChannel.Result) {
        result.success(CAS.settings.interstitialInterval)
    }

    private fun setInterstitialInterval(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<Int>("delay", result) {
            CAS.settings.interstitialInterval = it
        }
    }

    private fun restartInterstitialInterval(result: MethodChannel.Result) {
        CAS.settings.restartInterstitialInterval()
        result.success()
    }

    private fun isAllowInterstitialAdsWhenVideoCostAreLower(result: MethodChannel.Result) {
        result.success(CAS.settings.allowInterstitialAdsWhenVideoCostAreLower)
    }

    private fun allowInterstitialAdsWhenVideoCostAreLower(
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        call.getArgAndReturn<Boolean>("enable", result) {
            CAS.settings.allowInterstitialAdsWhenVideoCostAreLower = it
        }
    }

    private fun getLoadingMode(result: MethodChannel.Result) {
        result.success(CAS.settings.loadingMode)
    }

    private fun setLoadingMode(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<Int>("loadingMode", result) {
            CAS.settings.loadingMode = it
        }
    }

}